Add “Component - Styling” content (#66)
WebComponentsGuide/webcomponents.guide: learn/components/styling.md

order: 5
---

Web Components have powerful styling capabilities which make them portable and extensible. Styles declared within the
Shadow DOM serve as a Web Component’s default styling. It’s similar to writing a user-agent stylesheet for your custom
element and it works quite similarily as well.

## Shadow Encapsulation: Scoped Styles

Styles declared outside of a shadow tree cannot directly select elements within a shadow tree. Conversely, styles
declared inside a shadow tree cannot directly select elements outside of a shadow tree. For both of these points, there
are special pseudo-selectors which allow styles to cross the boundary of the Shadow DOM boundary — and we will cover

> them later on — but first we’ll focus on shadow encapsulation and its effect on styling. Here’s a basic demonstration:

Author: I think that’s a great suggestion. I definitely agree with the hard-to-readness of the first two sentences despite their careful wording and symmetry.

  </template>
</fancy-p>
```

The `<p>` element within the shadow tree is not effected by the styles declared outside of the shadow tree.

## Inheritance

Custom elements abide by the same rules of inheritance as other HTML elements. Properties such as `font-size`,
`font-family`, and `color` are inherited from a parent element by default. So are [CSS custom properties][]. Top-level
elements within a shadow tree inherit their inheritable properties from the custom element itself (also known as the
shadow host). Here’s a demonstration of this at play:

```html
<style>
  article {
    color: deeppink;
  }
</style>

<article>
  <h1>This text is deeppink.</h1>
  <article-meta>
    <template shadowroot="open">
      <style>
        span {
          font-style: italic;
        }
      </style>
      <span>By Some Person</span>
    </template>
  </article-meta>
</article>
```

The `<article-meta>` custom element inherits its `color` from the `<article>` element where it is set to `deeppink`. The
`<span>` element within the shadow tree inherits its `color` from the `<article-meta>` custom element which means the
value will be `deeppink`.

## Styling elements outside of a shadow tree

> In order to be portable, Web Components can provide default styles for a couple types of elements which exist outside of
> its shadow tree: the custom element itself (also known as the shadow host) and slotted elements.

Reviewer: I think this paragraph needs some refinement. I feel uneasy about "couple types of elements". Perhaps splitting the two concepts would work better... something like:

```suggestion
In order to be portable, Web Components can provide default styles for the element itself (also known as the shadow host).
They can also style slotted elements with some default styles.
```

### Writing default styles for the shadow host with `:host` and `:host()`

There are two selectors which can be used to style the shadow host from within the shadow tree. They are the `:host`
pseudo-class and the `:host()` pseudo-class function. The first will always select the shadow host. Here’s `:host` in

> action:

Reviewer: Front-loading knowledge of both selector types here is really nice :ok_hand:

```html
<fancy-p>
  <template shadowroot="open">
    <style>
      :host {
        display: inline-block;
      }
    </style>
  </template>
</fancy-p>
```

> The second will select the shadow host if it matches a selector. For example, given the following default styles for the
> `<fancy-p>` component, only the second element in the subsequent code example will have the styles applied.

Reviewer: This paragraph feels a bit too contingent on surrounding context. Perhaps:

```suggestion
The `:host()` selector will select the shadow host if it matches a given selector. For example, it can be  used to
select for hosts that have a given attribute. While `:host` may apply to `<fancy-p>` component, `:host([extra])` would
apply only to `<fancy-p extra>` elements:
```

```css
:host([extra]) {
  font-style: italic;
  font-weight: bold;
}
```

```html
<fancy-p>I not am extra</fancy-p>
<fancy-p extra>I am extra</fancy-p>
```

#### Chaining selectors after `:host`

While the `:host` selector refers to the shadow host element which is outside of the shadow tree, if you chain selectors
it will select elements within the shadow tree.

```html
<fancy-p>
  <template shadowroot="open">
    <style>
      :host > p {
        color: deeppink;
      }
    </style>
    <p>I am deeppink.</p>
  </template>
  <p>I am not deeppink.</p>
</fancy-p>
```

> You can `::slotted()` to style light DOM descendants of a shadow host.

Reviewer: This feels like it's not relevant to the heading here. Maybe this can be removed?

```suggestion
```

> ### Writing default styles for slotted elements with `::slotted()`

Reviewer: One thing missing from this section is `::slotted(*)` which feels like something worth documenting.

> The `::slotted()` pseudo-element selector allows you to write default styles for different kinds of elements which are
> slotted.

Reviewer: "different kinds of elements" feels a little ambiguous to me. I wonder if this works:

```suggestion
The `::slotted()` pseudo-element selector allows you to write default styles for slotted elements that match the given
selector. Specifying a selector can be useful for styling specific elements in particular ways.
```

```html
<fancy-elements>
  <template shadowroot="open">
    <style>
      ::slotted(button) {
        color: deeppink;
      }

      ::slotted(p) {
        color: teal;
      }
    </style>
    <slot></slot>
  </template>
  <button>I am a slotted button</button>
  <p>I am a slotted paragraph.</p>
</fancy-elements>
```

If you want to target elements in specific slots you can pass an attribute selector which matches the slot:

```html
<fancy-article>
  <template shadowroot="open">
    <style>
      ::slotted([slot="title"]) {
        font-size: 2rem;
      }
      ::slotted([slot="subtitle"]) {
        font-size: 1.25rem;
      }
      /* The following will target elements going into the unnamed slot */
      ::slotted(:not([slot])) {
        color: deeppink;
      }
    </style>
    <article>
      <hgroup>
        <slot name="title"></slot>
        <slot name="subtitle"></slot>
      </hgroup>
      <slot></slot>
    </article>
  </template>
  <h1 slot="title">I am the title</h1>
  <p slot="subtitle">I am the subtitle</p>
  <p>I am content.</p>
</fancy-article>
```

> You cannot chain selectors with `::slotted()`, so the following will not work:

Reviewer: This feels important enough to have a subheading, I feel. Also, I wonder if we can present alternative options. How does one overcome this barrier (more custom elements :smiley:)?

```css
::slotted(h1) span {
  color: deeppink;
}
```

> ## Parts: styling a shadow tree from the outside

Reviewer: I definitely think Parts deserve their own section, especially as by this point I think we're close to 10 minutes of reading time.

Author: Here’s a thought I just had for how to structure all the styling content maybe in different section:

- Introduction
  - Focuses on encapsulation and then gives a brief overview with links to each of the following sections.

- Ways to include styles in the Shadow DOM
  - A necessary but boring prerequisite for everything.

- How to write default styles for Web Components.
  - Covers `:host` and `::slotted`.
  - Shows how to make Web Component Styling portable.

- How to create a styling API for Web Components.
  - Covers using custom properties, `::part()`, `exportparts` for nested custom elements, and maybe even `!important` to limit styling certain properties.
  - Shows how to make Web Component Styling extensible.

There is more to cover like gotchas around `@keyframes` and `@font-face`. Maybe those could go in the “default styles” section.

Reviewer: I really like it. I like the idea of layering knowledge. What's the first thing that someone needs to just get styles working? Armed with that knowledge up front, they can experiment, and learn things independently, rather than having to read the entirety. Perhaps the basics can be expanded upon. Talk more about how it could be preferable to simplify selectors given the encapsulation. Maybe the introduction and "ways to include..." can be combined into a "Here's the basics" section.

Parts, while foundational, are more about architecting a component. So we should batch that with other architectural parts. So yeah I think having a section about styling API & extensibility is great.

The CSS Shadow Part API allows elements within a shadow tree to be styled from outside of it. This allows Web Components
to be very extensible.

```css
fancy-article::part(header) {
  display: grid;
  gap: 0.25rem;
  padding: 0.5rem;
  border-block-end: 0.125rem solid deeppink;
}

fancy-article::part(content) {
  display: grid;
  justify-content: start;
  gap: 0.5rem;
}
```

```html
<fancy-article>
  <template shadowroot="open">
    <article part="article">
      <hgroup part="header">
        <slot name="title"></slot>
        <slot name="subtitle"></slot>
      </hgroup>
      <div part="content">
        <slot name="content"></slot>
      </div>
    </article>
  </template>
  <h1 slot="title">Hello, World!</h1>
  <p slot="subtitle">I am a subtitle...</p>
  <p>I am content</p>
</fancy-article>
```

Similar to `::slotted()`, you cannot chain selectors after `::part()` to select children or siblings. The following will
not work:

```css
fancy-article::part(header) slot {
  display: block;
}
```

> ## How to include default styles for a Web Component

Reviewer: It might be worth lifting this section way up to the top of the document.

There are a variety of methods to include styles for a Web Component. Some will be familiar, but others are newer.

### Using `<style>`

The `<style>` tag is the most simple way to write styles for a Web Component. Just include it in your shadow tree:

```html
<style>
  :host {
    color: deeppink;
  }
</style>
```

### Using `<link rel="stylesheet">`

Using a `<link rel="stylesheet">` element in the shadow tree will allow you to write styles in an external stylesheet.

```html
<link rel="stylesheet" href="/fancy-article-element.css">
```

If you do this, the stylesheet will be loaded after the script is loaded. This will likely cause a “flash of unstyled
content” (FOUC). To circumvent this, you can preload the stylesheet like this:

```html
<link rel="preload" href="/fancy-article-element.css" as="style">

> ```

Reviewer: This would need to be in `<head>` right?

Author: Not necessary, technically it could be anywhere since it isn’t rendered, but on page load `<head>` would be the best place. A scenario where it wouldn’t really matter if it were in the `<head>` would be if a web component was dynamically imported later on. That’s a more advanced use case though. The thing with this suggestion in general is that the preload and the usage of the component on the page can’t happen at different times, otherwise, the preload is worthless and you’ll get a console warning about it. I think maybe I can rework this suggestion to emphasize the caution in doing this. Generally I think `<style>` and constructable stylesheets seem better because of this.
